Sparse bitmask

tpietzsch opened the pull request:
`SparseBitmask` is a `RandomAccessible<BoolType>` based on *n*Tree (quadtree, octree, ...) with little bitmasks (basically `ArrayImg<BitType>`) in the leafs. It's a `RandomAccessible` (not `Interval`). The tree grows on demand, just set pixels wherever...

There are efficient algorithms for
 * computation of bounding box of set (`true`) pixels,
 * iteration of set (`true`) pixels.

These are provided through a (read-only) view as an `IterableRegion`.

Some usage examples are in src/test/.

![sparsetree](https://user-images.githubusercontent.com/622070/50832760-7521bc00-134f-11e9-9a36-b062e02c5524.gif)

I'm happy with the underlying tree and bitmask stuff (`SparseBitmaskNTree`, `Tree`, `GrowableTree`).
However, packaging as a `RandomAccessible` in `SparseBitmask` required some trade-offs that are not so clear and depend on use case. I would be happy about some input...

In particular, read/write of pixels is protected by a `ReentrantReadWriteLock` that is acquired per pixel access. It would be nice to have this more coarse-grained, but I don't know how.

Also, `SparseBitmask.region()` provides `IterableRegion` view of the current state, and will throw `ConcurrentModificationException` when the bitmask is modified. There are some minor trade-offs there as well, but read/write locking is the major problem.

tpietzsch committed:
A bitmask stored as a quadtree (or n-dimensional equivalent).

The tree is truncated such that it does not contain boolean value-nodes
all the way to the leafs. Instead, leafs store bitmasks of a specified
size.

Also contains algorithms to efficiently compute the number and bounding box
of set pixels in such bitmasks.

tpietzsch committed:
Add SparseBitmask:

A RandomAccessible<NativeBoolType> that grows on demand.
SparseBitmask.tree() provides the underlying SparseBitmaskNTree.
SparseBitmask.region() provides a view as IterableRegion, with efficient
computation of bounding box and iteration of set pixels.

tpietzsch committed:
Add SparseBitmask usage examples.

imagejan commented on pom.xml:
```diff
@@ -197,5 +197,10 @@ Jean-Yves Tinevez and Michael Zinsmaier.</license.copyrightOwners>
 		</dependency>
+		<dependency>
+			<groupId>sc.fiji</groupId>
+			<artifactId>bigdataviewer-vistools</artifactId>
```
The Maven build on Travis now fails because `bigdataviewer-vistools` contains a transient dependency on `imglib2-roi`, creating a circular dependency.

As this is a `test` scope dependency only, the issue can be avoided by adding an exclusion:

```
		<dependency>
			<groupId>sc.fiji</groupId>
			<artifactId>bigdataviewer-vistools</artifactId>
			<scope>test</scope>
			<exclusions>
				<exclusion>
					<groupId>net.imglib2</groupId>
					<artifactId>imglib2-roi</artifactId>
				</exclusion>
			</exclusions>
		</dependency>
```

I'm not sure though if it is the best way to address this.

ctrueden replied:
Interesting hack—I didn't know that worked.Personally, I'd vote for the usage examples requiring vistools to move downstream. Perhaps into vistools itself, since it already depends on imglib2-roi transitively.

maarzt committed:
Make Bitmask a outer class

The functionality of Bitmask is very much independent of the class Tree.
It should be an independent class.

maarzt committed:
WIP: add tiny tests + THERE'S A BUG

Only on bit of SparseBitmask is true.
But SparseBitmask.region().cursor() seems to have two elements?

maarzt commented:
@tpietzsch I made some quick changes.
1. Bitmask can be a separate class. That should make the class Tree more readable.
2. There's a bug with the SpareIterable.region().cursor(), if there one bit set, it iterates over that pixel twice? So there probably a small bug in hasNext.

tpietzsch commented on src/main/java/net/imglib2/roi/sparse/SparseBitmaskNTree.java:
```diff
@@ -45,7 +42,9 @@
-	 * created. If possible, nodes will be merged.
+	 * created. If possible, nodeswithin bounds of this tree, i.e., within the
+	 * interval covered by the root node.
+	 will be merged.
```
Something went wrong here. The javadoc comments got garbled

tpietzsch replied:
I'll fix it...

tpietzsch committed:
"Fix" circular test dependency. Thanks @imagejan for this solution!!!

tpietzsch committed:
bugfix

tpietzsch commented:
@maarzt I fixed the bug. Thanks for noticing!

maarzt commented:
Bitmask is probably not a got term to refer to a black/white image. Mask or Bitmap are preferable. Bitmask are rather one dimensional.
https://en.wikipedia.org/wiki/Bitmap
https://en.wikipedia.org/wiki/Mask

maarzt committed:
Renamings:
* numSet -> numberOfOnes (numSet is too ambiguous. Is it a set of numbers?)
* leafDims -> dims
* bytes -> data (The javadoc comment suggests this renaming.)

Add method byteMask similar to byteIndex to reduce code duplication.

maarzt committed:
Rename bbox to boundingBox

maarzt committed:
Add FIXME comment for DefaultInterval

maarzt committed:
Important to avoid wrong usage of this method

maarzt committed:
TODO comment implement Intervals.wrap()

maarzt committed:
Rename BitMask to LeafBitmask

The design of this class is specific to it's use in Tree this
should be reflected by the name

maarzt committed:
Tree: introduce method getChildIndex

maarzt committed:
Add Benchmark to compare performance of NTree agains SparseBitmask

maarzt committed:
Add Benchmark against Labkit's sparse bitmaps.

The sparse bitmaps in Labkit use list's of coordinates to store only the
white pixels.

maarzt committed:
Tree: move operations that modify NodeData into NodeData

maarzt committed:
Make NodeData an outer class

maarzt committed:
NodeData: rename data to bitmask

maarzt committed:
Add final

maarzt committed:
Improve thread-safety of Tree.get and Tree.getNode

If a set operation sets NodeData.children or NodeData.bitmask to zero,
get will return NodeData.value.

maarzt committed:
Make tree a tree with fixed height

That way GrowableTree and Tree can be better distinguished.
Also reading Tree becomes thread-safe. The tree can always be red in any
thread, only the write operations need to be synchronized.

maarzt committed:
WIP

maarzt committed:
GrowableTree: some renamings

maarzt committed:
Make GrowableTree thread safe

maarzt committed:
Add read benchmark

maarzt committed:
Remove ReentrantReadWriteLock

This has a huge performance improvement for cuncurrent reading.
